src/librustc_typeck/check/expr.rs

```diff
@@ -159,13 +174,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let ty = self.check_expr_kind(expr, expected, needs);
 
         // Warn for non-block expressions with diverging children.
-        match expr.node {
-            ExprKind::Block(..) | ExprKind::Loop(..) | ExprKind::Match(..) => {},
-            ExprKind::Call(ref callee, _) =>
-                self.warn_if_unreachable(expr.hir_id, callee.span, "call"),
-            ExprKind::MethodCall(_, ref span, _) =>
-                self.warn_if_unreachable(expr.hir_id, *span, "call"),
-            _ => self.warn_if_unreachable(expr.hir_id, expr.span, "expression"),
+        if !is_try_block_ok_wrapped_expr {
```

Comment:
Use a guard on `ExprKind::Call` instead?

Reply:
Do you mean something like
```rust
match expr.node {
    ...
    ExprKind::Call(ref callee, _) if !expr.span.is_desugaring(DesugaringKind::TryBlock) =>
        self.warn_if_unreachable(expr.hir_id, callee.span, "call"),
    ...
}
```
?

Reply:
Yea, that or `... if !is_try_block_ok_wrapped_expr =>`.

Reply:
Well that didn't quite work because of the catch-all arm :) But it can be done with a separate arm.